allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -209,39 +220,142 @@ def from_params(cls, vocab, params: Params) -> 'WikiTablesSemanticParser':
                    attention_function=attention_function)
 
 
-class WikiTablesDecoderState(DecoderState):
+# This syntax is pretty weird and ugly, but it's necessary to make mypy happy with the API that
+# we've defined.  We're using generics to make the types of `combine_states` and `split_finished`
+# come out right.  See the note in `nn.decoding.decoder_state.py` for a little more detail.
+class WikiTablesDecoderState(DecoderState['WikiTablesDecoderState']):
     """
     Parameters
     ----------
-    action_history : ``List[int]``
-        The list of actions taken so far in this state.
-
-        The type annotation says this is an ``int``, but none of the training logic relies on this
-        being an ``int``.  In some cases, items from this list will get passed as inputs to
-        ``DecodeStep``, so this must return items that are compatible with inputs to your
-        ``DecodeStep`` class.
+    batch_indices : ``List[int]``
+        Passed to super class; see docs there.
+    action_history : ``List[List[int]]``
+        Passed to super class; see docs there.
+    score : ``List[Variable]``
+        Passed to super class; see docs there.
+    hidden_state : ``List[Variable]``
```

Comment:
We should keep these type annotations as `Tensors` because:
1. They provide more information to the user.
2. Pytorch will eventually remove `Variables`.
3. We have `Tensors` everywhere in the codebase where they are `Variables`.
4. With the exception of a few places where we catch these errors, like in `Metrics` everything is implicitly a `Variable` anyway.

Reply:
I'm not sure I agree with this.  If I have a `Variable`, that means I need to call `.data` to get the tensor.  I think having the type annotated as `Tensor` when I actually have a `Variable` is confusing.  If our goal is that the user knows the data type of the underlying tensor, we can easily solve that with a docstring comment.

Do you have any more info on the eventual union of `Tensor` and `Variable` in pytorch?  Either way, until that happens, I'd rather have our code actually have correct types, and who knows how long it'll actually take for the pytorch folks to merge them, and what it will look like when they do.

Reply:
What's more confusing is if we aren't consistent about this - everywhere apart from here we use tensor annotations, so I think you should make this consistent and if you feel strongly, change the annotations across the codebase. In 90% of cases, you don't have to think about whether a tensor is a variable. I don't know more about when the change is going to come, however, but at least in this PR I think you should be consistent, given that it is strictly worse to have a mix of tensor and variable types scattered throughout allennlp.

Reply:
I'd prefer to move all of our type annotations to `Variable` now, in places where the type actually is a variable.  @joelgrus did this in #374 - should we merge that PR, or part of it?  Do either of you have any objections to switching to `Variable` and adding data type info to docstrings?